File: edu_python/edu.py
import hashlib
import re
import mysql.connector
import requests
from bs4 import BeautifulSoup
from mail import mail


class edu(object):
    # 固定参数
    __schoolcode = '10611'
    # 登录链接
    __login_url = 'http://222.198.128.126/_data/index_login.aspx'
    # 成绩链接
    __score_url = 'http://222.198.128.126/xscj/Stu_MyScore_rpt.aspx'
    # 成绩分布
    __score_dis_url = 'http://222.198.128.126/xscj/Stu_cjfb_rpt.aspx'
    # 考试安排
    __exam_url = 'http://222.198.128.126/KSSW/stu_ksap_rpt.aspx'
    # 考表
    __exam_table_url = 'http://222.198.128.126/KSSW/stu_ksap_rpt.aspx'
    # 获取成绩参数
    score_data = {
        'sel_xn': 2015,  # 学年
        'sel_xq': 1,  # 学期
        'SJ': 0,  # 0为原始成绩，1为有效成绩
        'btn_search': '检索',
        'SelXNXQ': 2,
        'zfx_flag': 0,
    }
    # 获取考表参数
    exam_data = {
        'sel_xnxq': '20151',
        'sel_lc': '2015104,2015-2016学年第二学期13-16周',
        'btn_search': '检索',
    }
    # 获取成绩分布参数
    score_dis_data = {
        'sel_xn': '2015',
        'sel_xq': '0',
        'SelXNXQ': '2',
        'submit': '检索',
    }
    # 考表参数
    exam_table_data = {
        'sel_xnxq': '20151',
        'sel_lc': '2015104,2015-2016学年第二学期13-16周',
        'btn_search': '检索',
    }

    def __init__(self, code, password):
        super(edu, self).__init__()
        self.__code = code
        self.__pwd = password
        self.login()

    # 得到view参数

    def __getView(self):
        view = []
        r = re.compile(
            r'<input type="hidden" name="__VIEWSTATE" value="(.*?)" \/>')
        data = requests.get(self.__login_url)
        view = r.findall(data.text)
        r = re.compile(
            r'<input type="hidden" name="__VIEWSTATEGENERATOR" value="(.*?)" \/>')
        data = r.findall(data.text)
        view.append(data[0])
        return view

    # 加密函数

    def __checkPwd(self):
        p = hashlib.md5(self.__pwd.encode()).hexdigest()
        p = hashlib.md5(
            (self.__code + p[0:30].upper() + self.__schoolcode).encode()).hexdigest()
        return p[0:30].upper()

    # 登录获取cookie

    def login(self):
        view = self.__getView()
        params = {'__VIEWSTATE': view[0],
                  '__VIEWSTATEGENERATOR': view[1],
                  'Sel_Type': 'STU',
                  'txt_dsdsdsdjkjkjc': self.__code,
                  'efdfdfuuyyuuckjg': self.__checkPwd(), }
        r = requests.post(self.__login_url, data=params)
        if u'正在加载权限数据' in r.text:
            print('登录成功！')
            print('学号为：' + self.__code + ',密码为：' + self.__pwd)
            self.__cookies = r.cookies
            return True

    # post获取数据

    def __post(self, url, data):
        return requests.post(
            url, data, cookies=self.__cookies).text

    # 获取数据

    def catchData(self, type):
        result = {'score': self.__post(self.__score_url, self.score_data), 'exam': self.__post(self.__exam_url, self.exam_data), 'score_dis': self.__post(
            self.__score_dis_url, self.score_dis_data), 'exam_table': self.__post(self.__exam_table_url, self.exam_table_data), }
        return result.get(type)

    # 插入数据库

    def insert_score(self, X):
        conn = mysql.connector.connect(
            host='localhost', user='root', password='113789', database='mail')
        cur = conn.cursor()
        for tr in X:
            d = []
            for td in tr:
                d.append(td.text)
            sql = "select * from score where id='%s'" % d[0]
            cur.execute(sql)
            a = cur.fetchall()
            if not a:
                sql = "insert into score(id,course,credit,type,method,quality,score,mark) values('%s','%s','%s','%s','%s','%s','%s','%s')" % (
                    d[0], d[1], d[2], d[3], d[4], d[5], d[6], d[7])
                cur.execute(sql)
                conn.commit()
            else:
                if not d[6] in a[0]:
                    sql = "update score set score='%s' where id='%s'" % (d[6], d[
                        0])
                    cur.execute(sql)
                    conn.commit()
                    email = mail()
                    email.textMail('******', '有成绩更新了', tr)

        cur.close()
        conn.close()
    # 爬取成绩并存入数据库

    def get_score(self):
        data = self.catchData('score')
        bs4 = BeautifulSoup(data, 'lxml')
        B = bs4.find_all(class_='B')
        H = bs4.find_all(class_='H')
        self.insert_score(B)
        self.insert_score(H)

my_edu = edu('****', '*****')

File: edu_php/edu.php
<?php
require 'simple_html_dom.php';
class Edu {
    //cookie保存路径
    private $cookie;
    //学号
    private $code;
    //密码
    private $pwd;
    //固定参数
    private $schoolcode = '10611';
    //登录链接
    private $login_url = 'http://222.198.128.126/_data/index_login.aspx';
    //成绩链接
    private $score_url = 'http://222.198.128.126/xscj/Stu_MyScore_rpt.aspx';
    //成绩分布
    private $score_dis_url = 'http://222.198.128.126/xscj/Stu_cjfb_rpt.aspx';
    //考试安排
    private $exam_url = 'http://222.198.128.126/KSSW/stu_ksap_rpt.aspx';
    //考表
    private $exam_table_url = 'http://222.198.128.126/KSSW/stu_ksap_rpt.aspx';
    //考试安排页面
    private $exam_plan_url = 'http://222.198.128.126/KSSW/stu_ksap.aspx';
    //获取成绩参数
    public $score_data = array(
        'sel_xn'     => 2015, //学年
        'sel_xq'     => 0, //学期
        'SJ'         => 1,
        'btn_search' => '检索',
        'SelXNXQ'    => 2,
        'zfx_flag'   => 0,
    );
    //获取考表参数
    public $exam_data = array(
        'sel_xnxq'   => '20151',
        'sel_lc'     => '2015104,2015-2016学年第二学期13-16周',
        'btn_search' => '检索',
    );
    //获取成绩分布参数
    public $score_dis_data = array(
        'sel_xn'  => '2015',
        'sel_xq'  => '0',
        'SelXNXQ' => '2',
        'submit'  => '检索',
    );
    //考表参数
    public $exam_table_data = array(
        'sel_xnxq'   => '20151',
        'sel_lc'     => '2015104,2015-2016学年第二学期13-16周',
        'btn_search' => '检索',
    );
    //select标签所需参数
    private $select_data = array('vT' => 'stu', 'vP' => 'xnxqkslc');
    //构造函数
    public function __construct($schoolcode_, $password) {
        $this->code = $schoolcode_;
        $this->pwd  = $password;
        if (!file_exists($this->cookie)) {
            $data = $this->login();
        }
    }
    //得到select所需参数
    private function getSelectData() {
        $url    = $this->exam_plan_url;
        $result = $this->get($url);
        $html   = str_get_html($result);
        $wd     = $html->find('#thexnxqkslc');
        $id     = $html->find('option');
        foreach ($wd as $value) {
            $this->select_data['wd'] = $value->width;
        }
        foreach ($id as $value) {
            $this->select_data['id'] = $value->value;
        }
    }
    //得到考表select标签
    public function getTableSelect() {
        $this->getSelectData();
        $url  = 'http://222.198.128.126/KSSW/Private/list_xnxqkslc.aspx?id=' . $this->select_data['id'] . '&wd=' . $this->select_data['wd'] . '&vP=' . $this->select_data['vP'] . '&vT=' . $this->select_data['vT'];
        $data = iconv('GBK', 'UTF-8//IGNORE', $this->get($url));
        preg_match('/<select.*select>/', $data, $matches);
        return $matches[0];
    }
    //得到成绩select标签
    public function getScoreSelect() {
        $url    = 'http://222.198.128.126/xscj/Stu_MyScore.aspx';
        $data   = $this->get($url);
        $data   = iconv('GBK', 'UTF-8//IGNORE', $data);
        $html   = str_get_html($data);
        $result = $html->find('form');
        return $result;
    }
    //得到view参数
    private function getView() {
        $url     = $this->login_url;
        $result  = $this->get($url);
        $pattern = '/<input type="hidden" name="__VIEWSTATE" value="(.*?)" \/>/is';
        preg_match_all($pattern, $result, $matches);
        $res[0]  = $matches[1][0];
        $pattern = '/<input type="hidden" name="__VIEWSTATEGENERATOR" value="(.*?)" \/>/is';
        preg_match_all($pattern, $result, $matches);
        $res[1] = $matches[1][0];
        return $res;
    }
    //post方法得到cookie
    private function post($url, $post_data) {
        $this->cookie = dirname(__FILE__) . '\cookie_edu.txt';
        $ch           = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_HEADER, 0);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_POST, 1);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($ch, CURLOPT_COOKIEJAR, $this->cookie);
        curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($post_data));
        $result = curl_exec($ch);
        curl_close($ch);
        return $result;
    }
    //get方法
    private function get($url) {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_HEADER, 0);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($ch, CURLOPT_COOKIEFILE, $this->cookie);
        $result = curl_exec($ch);
        return $result;
    }
    //post方法得到数据
    private function postData($url, $post_data) {
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_USERAGENT, 'User-Agent:Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/38.0.2125.122 Safari/537.36 SE 2.X MetaSr 1.0');
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_HEADER, 0);
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_POST, 1);
        curl_setopt($ch, CURLOPT_COOKIEFILE, $this->cookie);
        curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($post_data));
        $data = curl_exec($ch);
        curl_close($ch);
        return $data;
    }
    //密码加密函数
    private function checkPwd($code, $pwd, $schoolcode) {
        return strtoupper(substr(md5(($code . strtoupper(substr(md5($pwd), 0, 30)) . $schoolcode)), 0, 30));
    }
    //登录获取cookie
    private function login() {
        $view      = $this->getView();
        $post_data = array(
            '__VIEWSTATE'          => $view[0],
            '__VIEWSTATEGENERATOR' => $view[1],
            'Sel_Type'             => 'STU',
            'txt_dsdsdsdjkjkjc'    => $this->code,
            'efdfdfuuyyuuckjg'     => $this->checkPwd($this->code, $this->pwd, $this->schoolcode),
        );
        $url    = $this->login_url;
        $data   = iconv('GBK', 'UTF-8//IGNORE', $this->post($url, $post_data));
        $result = strpos($data, '正在加载权限数据');
        return $result;
    }
    //获取各类数据方法
    public function catchData($type) {
        switch ($type) {
            case 'score':
                $url       = $this->score_url;
                $post_data = $this->score_data;
                break;
            case 'exam':
                $url       = $this->exam_url;
                $post_data = $this->exam_data;
                break;
            case 'score_dis':
                $url       = $this->score_dis_url;
                $post_data = $this->score_dis_data;
                break;
            case 'exam_table':
                $url       = $this->exam_table_url;
                $post_data = $this->exam_table_data;
                break;
            default:
                break;
        }
        $data = iconv('GBK', 'UTF-8//IGNORE', $this->postData($url, $post_data));
        return $data;
    }
}
$my_edu = new Edu('******', '*****');
$data   = $my_edu->catchData('exam');